File: 2019/utils/timings.go
package main

import (
	"fmt"
	"time"
)

func formatDuration(duration time.Duration) (float64, string) {
	seconds := duration.Seconds()

	switch {
	case seconds < 1:
		return seconds * 1000, "ms"
	case seconds < 60:
		return seconds, "s"
	case seconds < 3600:
		return seconds / 60, "min"
	default:
		return seconds / 3600, "h"
	}
}

func runFunc(name string, f func() int) {
	start := time.Now()
	result := f()
	end := time.Now()
	elapsed := end.Sub(start)

	duration, unit := formatDuration(elapsed)

	// ANSI escape code for setting text color to gray
	gray := "\033[90m"

	// ANSI escape code for resetting text color
	reset := "\033[0m"

	fmt.Printf("%s: %d %s(%.2f %s)%s\n", name, result, gray, duration, unit, reset)
}

File: 2019/utils/timings.py
from time import perf_counter
from typing import Callable, Tuple, TypeVar

from termcolor import colored

T = TypeVar("T")


def format_duration(milliseconds: float) -> str:
    if milliseconds < 1000:
        return f"{milliseconds:.3f}ms"
    elif milliseconds < 60 * 1000:
        return f"{milliseconds / 1000:.3f}s"
    elif milliseconds < 60 * 60 * 1000:
        return f"{milliseconds / (60 * 1000):.3f}min"
    else:
        return f"{milliseconds / (60 * 60 * 1000):.3f}h"


def profile_run(name: str, func: Callable[[], T]) -> Tuple[T, float]:
    start = perf_counter()
    result = func()
    end = perf_counter()
    duration = (end - start) * 1000

    print(
        f"{colored(name, 'magenta')}: {colored(result, 'green')} {colored(f'(executed in {format_duration(duration)})', 'light_grey', attrs=['dark'])}"
    )

    return result, duration
